From a5cda44b9d17cfae5931c4efe66d65bbdcbfbfda Mon Sep 17 00:00:00 2001
From: Damian Peckett <damian@pecke.tt>
Date: Fri, 18 Oct 2024 12:37:00 +0200
Subject: [PATCH] feat: add SOPHGO compatible boot header

So that u-boot can be loaded by the SOPHGO FSBL bootloader, we need to add a
header to the u-boot binary.

Upstream-Status: Inappropriate [SOPHGO specific]
Signed-off-by: Damian Peckett <damian@pecke.tt>
---
 arch/riscv/cpu/start.S         |  2 ++
 arch/riscv/include/asm/boot0.h | 35 ++++++++++++++++++++++++++++++++++
 2 files changed, 37 insertions(+)
 create mode 100644 arch/riscv/include/asm/boot0.h

diff --git a/arch/riscv/cpu/start.S b/arch/riscv/cpu/start.S
index 8e58f641f1..a0eb7dc95e 100644
--- a/arch/riscv/cpu/start.S
+++ b/arch/riscv/cpu/start.S
@@ -39,6 +39,8 @@ secondary_harts_relocation_error:
 .section .text
 .globl _start
 _start:
+#include <asm/boot0.h>
+_start_real:
 #if CONFIG_IS_ENABLED(RISCV_MMODE)
 	csrr	a0, CSR_MHARTID
 #endif
diff --git a/arch/riscv/include/asm/boot0.h b/arch/riscv/include/asm/boot0.h
new file mode 100644
index 0000000000..417a05af00
--- /dev/null
+++ b/arch/riscv/include/asm/boot0.h
@@ -0,0 +1,35 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * SOPHGO u-boot header
+ */
+
+#ifndef __BOOT0_H__
+#define __BOOT0_H__
+
+/* BOOT0 header information */
+	j boot0_time_recode
+	.balign 4
+	.word 0x33334c42  /* b'BL33' */
+	.word 0xdeadbeea  /* CKSUM */
+	.word 0xdeadbeeb  /* SIZE */
+#ifdef CONFIG_SPL_BUILD
+	.quad	CONFIG_SPL_TEXT_BASE
+#else
+	.quad   CONFIG_TEXT_BASE
+#endif
+	.word 0xdeadbeec
+	.balign 4
+	j boot0_time_recode
+	.balign 4
+/* BOOT0 header end */
+boot0_time_recode:
+	csrr x1, time
+	la x2, BOOT0_START_TIME
+	sw x1, 0(x2)
+	j _start_real
+
+	.global BOOT0_START_TIME
+BOOT0_START_TIME:
+	.word 0
+
+#endif /* __BOOT0_H__ */
-- 
2.39.5

